src/vcpkg/commands.build.cpp

```diff
@@ -1488,54 +1488,75 @@ namespace vcpkg
+        const auto max_log_length = static_cast<size_t>(max_size);
+        const auto start_block_length = max_log_length * 4 / 12;
+        const auto start_block_max_length = max_log_length * 5 / 12;
+        const auto end_block_length = max_log_length * 7 / 12;
+        const auto end_block_max_length = max_log_length - start_block_max_length;
+        auto log = fs.read_contents(path, VCPKG_LINE_INFO);
+        if (log.size() > max_log_length)
         {
-            static constexpr auto MAX_LOG_LENGTH = 50'000;
-            static constexpr auto START_BLOCK_LENGTH = 3'000;
-            static constexpr auto START_BLOCK_MAX_LENGTH = 5'000;
-            static constexpr auto END_BLOCK_LENGTH = 43'000;
-            static constexpr auto END_BLOCK_MAX_LENGTH = 45'000;
-            auto log = fs.read_contents(path, VCPKG_LINE_INFO);
-            if (log.size() > MAX_LOG_LENGTH)
+            auto first_block_end = log.find_first_of('\n', start_block_length);
+            if (first_block_end == std::string::npos || first_block_end > start_block_max_length)
             {
-                auto first_block_end = log.find_first_of('\n', START_BLOCK_LENGTH);
-                if (first_block_end == std::string::npos || first_block_end > START_BLOCK_MAX_LENGTH)
-                {
-                    first_block_end = START_BLOCK_LENGTH;
-                }
-
-                auto last_block_end = log.find_last_of('\n', log.size() - END_BLOCK_LENGTH);
-                if (last_block_end == std::string::npos || last_block_end < log.size() - END_BLOCK_MAX_LENGTH)
-                {
-                    last_block_end = log.size() - END_BLOCK_LENGTH;
-                }
-
-                auto first = log.begin() + first_block_end;
-                auto last = log.begin() + last_block_end;
-                auto skipped_lines = std::count(first, last, '\n');
-                log.replace(first, last, fmt::format("\n...\nSkipped {} lines\n...", skipped_lines));
+                first_block_end = start_block_max_length;
             }
 
-            while (!log.empty() && log.back() == '\n')
+            auto last_block_start = log.find_last_of('\n', log.size() - end_block_length);
+            if (last_block_start == std::string::npos || last_block_start < log.size() - end_block_max_length)
             {
-                log.pop_back();
+                last_block_start = log.size() - end_block_max_length;
             }
 
-            return fmt::format("<details><summary>{}</summary>\n\n```\n{}\n```\n</details>", path.native(), log);
+            auto first = log.begin() + first_block_end;
+            auto last = log.begin() + last_block_start;
+            auto skipped_lines = std::count(first, last, '\n');
+            out.append(log, 0, first_block_end);
+            fmt::format_to(std::back_inserter(out), "\n...\nSkipped {} lines\n...", skipped_lines);
+            out.append(log, last_block_start);
```

Comment:
```suggestion
            out.append(log.begin(), first);
            fmt::format_to(std::back_inserter(out), "\n...\nSkipped {} lines\n...", skipped_lines);
            out.append(last, log.end());
```

I think the iterator form makes it a bit clearer that the inserted content is talking about the same thing.